code/headers/hardware_usart.hpp

```diff
@@ -5,151 +5,254 @@
  * @brief Hardware implementation for the usart
  * @version 0.1
  * @date 2019-05-24
- * 
+ *
  * @copyright Copyright (c) 2019
- * 
+ *
  */
 
 #include <hwlib.hpp>
 #include <ringbuffer.hpp>
+#include <type_traits>
+
 #include <usart_connection.hpp>
 
 namespace r2d2::usart {
+    namespace detail::pio {
+        // helper structs for the peripherals
+        struct periph_a {};
+        struct periph_b {};
+
+        // helper structs for the pio's
+        struct pioa {};
+        struct piod {};
+        struct piob {};
+    }; // namespace detail::pio
 
-    enum class usart_ports {
-        uart1 = 0,
-        uart2,
-        uart3,
-        // only for determining how many uart ports there are
-        UART_SIZE
+    /**
+     * @brief Struct with the data for usart0.
```

Comment:
Nitpick: This isn't data but these are pin definitions., same goes for the other structs